### PR TITLE
Implement basic unnest function

### PR DESCRIPTION
**Pull request by jayzhan211**
# Which issue does this PR close?

Ref https://github.com/apache/arrow-datafusion/issues/6555

Closes #.

# Rationale for this change

Support `unnest` for arrays

# What changes are included in this PR?

- [x] `select unnest()`
- [X] `select * from unnest()`
- [X] unnest alias
- [X] unnest multidimensional arrays
- [X] unnest args more than 2
- [ ] select partial columns
- [ ] deterministic result for `datafusion.execution.target_partitions > 1` (We might not need this?)

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jayzhan211** [comment]
Current implementation.

There are two workflows for unnest
`select unnest()`, we will get statements from `select.projection`. `Unnest` is transformed to `Expr::Unnest`. We will then build the LogicalPlan in optimize phase.

`select * from unnest()`, we can get the table relation Unnest directly. In this case, we can build LogicalPlan without Expr::Unnest involved.

**Comment by jayzhan211** [comment]
Cleanup done, ready for review.
@izveigor @alamb

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 1]
```diff
@@ -531,3 +531,49 @@ SELECT
 FROM t
 ----
 true true
+statement ok
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:534]: This does not work in this PR, I plan to complete these in another PR. I will add it to the task list.

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 2]
```diff
@@ -531,3 +531,49 @@ SELECT
...
+set datafusion.execution.target_partitions = 1;
+
+query ?
+select unnest(make_array(1,2,3,4,5))
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:538]: This one also.

File: datafusion/sql/src/planner.rs [hunk 3]
```diff
@@ -228,12 +228,10 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         plan: LogicalPlan,
         alias: TableAlias,
     ) -> Result<LogicalPlan> {
-        let apply_name_plan = LogicalPlan::SubqueryAlias(SubqueryAlias::try_new(
-            plan,
-            self.normalizer.normalize(alias.name),
-        )?);
-
-        self.apply_expr_alias(apply_name_plan, alias.columns)
+        let plan = self.apply_expr_alias(plan, alias.columns)?;
```
Comment (jayzhan211) [inline datafusion/sql/src/planner.rs:231]: The order is changed so that we can get either `expr` or `table. expr`.

Example in array.slt
table.expr: `SELECT sum(data.a), sum(data.b) FROM unnest(make_array(1, 2, 3), make_array(4, 5, 6, 7)) as data(a, b);`
expr: `SELECT sum(a), sum(b) FROM unnest(make_array(1, 2, 3), make_array(4, 5, 6, 7)) as data(a, b);`

**Comment by jayzhan211** [comment]
Wait for sqlparser 0.36

**Comment by alamb** [comment]
I was planning to release sqlparser in a few weeks -- but if it is blocking this PR I can find time to make a release sooner

**Comment by jayzhan211** [comment]
> I was planning to release sqlparser in a few weeks -- but if it is blocking this PR I can find time to make a release sooner

I hope sqlparser can be released soon! Thanks!

**Comment by alamb** [comment]
Next release is tracked by https://github.com/sqlparser-rs/sqlparser-rs/issues/905

I will try and prioritize it but this week is short for me because of the US holiday (and I am spending most of my DataFusion time on #6800)

**Comment by jayzhan211** [comment]
@alamb Kindly remind you. :)

**Comment by jayzhan211** [comment]
Possible improvement:
Utilize the existing unnest_columns function instead building a new one.

TODO in another PR
- [ ] Support column-wise unnest
- [ ] Able to preserve nulls
- [ ] Select partial columns
- [ ] Subquery

**Comment by jayzhan211** [comment]
@izveigor @alamb Ready for review, thanks.

**Comment by jayzhan211** [comment]
@vincev @smiklos You might be interested about Unnest function.

**Comment by vincev** [comment]
I'll try to take a look tomorrow, yes I agree  maybe we can simplify the code if we can use `unnest_column` somehow.

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 4]
```diff
@@ -1794,9 +1802,150 @@ select make_array(f0) from fixed_size_list_array
...
+
+query IIIII
+select * from unnest(
+  make_array(1), 
+  make_array(2,3), 
+  make_array(4,5,6),
+  make_array(7,8),
+  make_array(9)
+);
+----
```
Comment (vincev) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:1854]: I find having `unnest` in the `from` clause a bit strange, [DuckDB](https://duckdb.org/docs/sql/query_syntax/unnest.html) syntax looks intuitive, the above would look like:

```
D SELECT * FROM (VALUES ([1]), ([2, 3]), ([4,5,6]), ([7,8]), ([9])) tbl(l);
┌───────────┐
│     l     │
│  int32[]  │
├───────────┤
│ [1]       │
│ [2, 3]    │
│ [4, 5, 6] │
│ [7, 8]    │
│ [9]       │
└───────────┘
```

and then unnesting:

```
D SELECT unnest(l) FROM (VALUES ([1]), ([2, 3]), ([4,5,6]), ([7,8]), ([9])) tbl(l);
┌───────────┐
│ unnest(l) │
│   int32   │
├───────────┤
│         1 │
│         2 │
│         3 │
│         4 │
│         5 │
│         6 │
│         7 │
│         8 │
│         9 │
└───────────┘
```

This would also work if we mix array and non array values:

```
D SELECT * FROM (VALUES ([11],1), ([21, 22],2), ([31,32,33],3), ([41,42],4), ([51],5)) tbl(l, m);
┌──────────────┬───────┐
│      l       │   m   │
│   int32[]    │ int32 │
├──────────────┼───────┤
│ [11]         │     1 │
│ [21, 22]     │     2 │
│ [31, 32, 33] │     3 │
│ [41, 42]     │     4 │
│ [51]         │     5 │
└──────────────┴───────┘
```

unnesting them:

```
D SELECT unnest(l), m FROM (VALUES ([11],1), ([21, 22],2), ([31,32,33],3), ([41,42],4), ([51],5)) tbl(l, m);
┌───────────┬───────┐
│ unnest(l) │   m   │
│   int32   │ int32 │
├───────────┼───────┤
│        11 │     1 │
│        21 │     2 │
│        22 │     2 │
│        31 │     3 │
│        32 │     3 │
│        33 │     3 │
│        41 │     4 │
│        42 │     4 │
│        51 │     5 │
└───────────┴───────┘
```

**Comment by vincev** [comment]
I had quick look through the code to see if we can apply `unnest_column`, if I understand correctly we create a `LogicalPlan` for each array column then apply a row number window function to each plan and then join all these plans with a `Full` join. 

I guess this makes sense when we create arrays in the select expression like:

```
select unnest(make_array(1,2,3,4,5)), unnest(make_array(6,7));
```

`unnest_column` is useful when we work on a full table that contains a arrays and non arrays like:

```
┌──────────────┬───────┐
│      l       │   m   │
│   int32[]    │ int32 │
├──────────────┼───────┤
│ [11]         │     1 │
│ [21, 22]     │     2 │
│ [31, 32, 33] │     3 │
│ [41, 42]     │     4 │
│ [51]         │     5 │
└──────────────┴───────┘
```

Then a statement like:

```
select unnest(l), m from table;
```

should call `unnest_column` on column `l` that will take take care of unnesting and null handling (see https://github.com/apache/arrow-datafusion/pull/7088).

File: datafusion/core/tests/sqllogictests/test_files/array.slt (hunk 4 above)
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:1854]: I also agree this syntax is better if we like partial unnest, e.g. select unnest(a), b, unnest(c) from table.
Clickhouse has the similar syntax https://clickhouse.com/docs/en/sql-reference/functions/array-join

Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:1854]: I think we can only keep `select * from unnest syntax`, since it is equivalent to `select unnest(a), unnest(b)...unnest(z) from table` but simplified

**Comment by alamb** [comment]
This looks awesome -- thank you @jayzhan211  -- I plan to take a look later today or tomorrow

File: datafusion/sql/tests/sql_integration.rs [hunk 5]
```diff
@@ -2962,36 +2962,37 @@ fn cte_with_no_column_names() {
...
+    Projection: Int64(1) AS a, Int64(2) AS b, Int64(3) AS c
+      Projection: Int64(1), Int64(2), Int64(3)
+        EmptyRelation"#;
 
-    quick_test(sql, expected)
+    quick_test(sql, expected);
 }
 
 #[test]
 fn cte_with_column_aliases_precedence() {
```
Comment (alamb) [inline datafusion/sql/tests/sql_integration.rs:2981]: are the changes in this test related to unnest? Maybe we could break them out into their own PR for review

**Review by alamb — COMMENTED** [review]
This PR is *super* cool @jayzhan211  -- thank you very much 🙏 

I had a few suggestions -- the main ones related to not panicing and returning `NotImplemented` and adding doc comments to public API functions. Otherwise this looks really great

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 6]
```diff
@@ -1794,9 +1802,150 @@ select make_array(f0) from fixed_size_list_array
 ----
 [[1, 2], [3, 4]]
 
+## Unnest
 
-### Delete tables
+# Set target partitions to 1 for deterministic results
+statement ok
+set datafusion.execution.target_partitions = 1;
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:1809]: `rowsort` might not be a good idea, if the result is in descending order, `rowsort` convert it to ascending forcibly, but if the code contains a bug that does not return in descending order, then we might miss it.

File: datafusion/optimizer/src/optimizer.rs [hunk 7]
```diff
@@ -422,6 +424,11 @@ fn assert_schema_is_the_same(
     prev_plan: &LogicalPlan,
     new_plan: &LogicalPlan,
 ) -> Result<()> {
+    if rule_name == "unnest_expressions" {
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:427]: There is the schema mismatch bug, I thought it is fine for `unnest`, let me take a look carefully.

File: datafusion/sql/tests/sql_integration.rs (hunk 5 above)
Comment (jayzhan211) [reply datafusion/sql/tests/sql_integration.rs:2981]: This change is not related to unnest directly, but they are affected by adding the `UnnestExpression` layer, so they are better changed together in this PR.

File: datafusion/optimizer/src/unnest_expressions.rs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+}
+
+impl UnnestExpressions {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn optimize_internal(plan: &LogicalPlan) -> Result<LogicalPlan> {
+        if let LogicalPlan::Projection(_) = plan {
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/unnest_expressions.rs:50]: Recursively visiting means there might have multiple layers of unnest expressions, like `unnest( unnest () )`. I think we might not need them, at least in this PR.

File: datafusion/optimizer/src/optimizer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:427]: External error: query failed: DataFusion error: unnest_expressions
caused by
Internal error: Optimizer rule 'unnest_expressions' failed, due to generate a different schema, original schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(1),Int64(2),Int64(3)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(4),Int64(5)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: "sq0" }), field: Field { name: "col0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: "sq1" }), field: Field { name: "col1", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
[SQL] select unnest(make_array(1,2,3)),
       unnest(make_array(4,5))
;

[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:427]: Two differences in the schema.
1. qualifier with sq0, sq1..., since I wrap each unnest column with subquery
2. column name, I rename them to col0, col1..., just for simplicity.

@alamb Does it makes sense to skip these two checks and only check the other?
If not, is there an existing API to change the schema for LogicalPlan, so I can set None to qualifier and create a new field with the original name?

[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:427]: let me check if those two are necessary or not.

[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:427]: Internal error: Optimizer rule 'unnest_expressions' failed, due to generate a different schema, original schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(1),Int64(2),Int64(3)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(4),Int64(5)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "make_array(Int64(1),Int64(2),Int64(3))", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "make_array(Int64(4),Int64(5),NULL)", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }

I don't know how to fix these schema changes

name changes: unnest(make_array(Int64(1),Int64(2),Int64(3))) -> make_array(Int64(1),Int64(2),Int64(3))
data_type changes: List(int64) -> int64

File: datafusion/expr/src/expr.rs [hunk 9]
```diff
@@ -931,6 +948,40 @@ impl Expr {
     pub fn contains_outer(&self) -> bool {
         !find_out_reference_exprs(self).is_empty()
     }
+
+    pub fn flatten(&self) -> Self {
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/expr/src/expr.rs:955]: Not exactly the opposite of unnest, we expect to get the 1D list of elements in the nested array in this function.

File: datafusion/expr/src/logical_plan/builder.rs [hunk 10]
```diff
@@ -1418,6 +1534,141 @@ pub fn unnest(input: LogicalPlan, column: Column) -> Result<LogicalPlan> {
     }))
 }
 
+// Create function name for unnest
+// Different from create_function_physical_name, so we create a customize one.
+// We need fun(arg1,arg2,arg3) format, but not fun(arg1, arg2, arg3).
+// TODO: We dont need this if we can customize format in impl::Display for Expr
+fn create_unnest_arguments_name(array_exprs: Vec<Expr>) -> Result<Vec<String>> {
```
Comment (jayzhan211) [inline datafusion/expr/src/logical_plan/builder.rs:1541]: New function to review

File: datafusion/optimizer/src/optimizer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:427]: Need to review again

**Comment by alamb** [comment]
I have this on my list for review, but may not get to it today

**Comment by alamb** [comment]
I will review this PR tomorrow, Monday

File: datafusion/expr/src/logical_plan/builder.rs [hunk 11]
```diff
@@ -1426,6 +1542,141 @@ pub fn unnest(input: LogicalPlan, column: Column) -> Result<LogicalPlan> {
...
+            }
+            e => Err(DataFusionError::NotImplemented(format!(
+                "Expr {e} is not supported in UNNEST"
+            ))),
+        })
+        .collect::<Result<Vec<String>>>()
+}
+
+/// Create unnest plan from arrays.
+fn build_unnest_plans(
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:1574]: I am surprised this can't use https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.LogicalPlan.html#variant.Unnest

**Review by alamb — COMMENTED** [review]
Thank you @jayzhan211  -- this is getting there but I don't think it is ready to merge yet (for reasons I left in comments below)

My major concerns are that we aren't using the `LogicalPlan::Unnest` and `UnnestExec` physical operator -- doing so would mean that improvements we plan (like making `UnnestExec` more efficient) will not apply to using `unnest` as a function

As this PR is large and I would like to move it along, I suggest
1. Remove `flatten` (we can put that into another PR)
2. Make a PR with just the `Expr::Unnest` -- maybe this could simply create a `LogicalPlan::Unnest`?
3. Make a PR to add any additional features needed for unnest (like multiple column support, for example)

What do you think?

cc @izveigor @vincev and @smiklos

File: datafusion/expr/src/expr.rs [hunk 12]
```diff
@@ -970,6 +987,43 @@ impl Expr {
...
     }
+
+    /// Flatten the nested array expressions until the base array is reached.
+    /// For example:
+    ///  unnest([[1, 2, 3], [4, 5, 6]]) => unnest([1, 2, 3, 4, 5, 6])
+    pub fn flatten(&self) -> Self {
+        self.try_flatten().unwrap_or_else(|_| self.clone())
+    }
+
+    pub fn try_flatten(&self) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/expr/src/expr.rs:998]: Since we only support flatten on array now so I check if it is make_array.

**Comment by jayzhan211** [comment]
> As this PR is large and I would like to move it along, I suggest
> 
> 1. Remove `flatten` (we can put that into another PR)
> 2. Make a PR with just the `Expr::Unnest` -- maybe this could simply create a `LogicalPlan::Unnest`?
> 3. Make a PR to add any additional features needed for unnest (like multiple column support, for example)

Let me work on https://github.com/apache/arrow-datafusion/issues/6995 first since I don't think simply creating LogicalPlan from Expr::Unnest is trivial.

**Comment by alamb** [comment]
Marking as draft to signify it is not waiting on review

File: datafusion/optimizer/src/optimizer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/optimizer/src/optimizer.rs:433]: https://github.com/apache/arrow-datafusion/blob/85ec31433615735a05d332f87cd0bdfc11aac663/datafusion/core/src/physical_plan/projection.rs#L423-L440

All nulls to align the row size at L432.

**Comment by jayzhan211** [comment]
> 2. Make a PR with just the Expr::Unnest -- maybe this could simply create a LogicalPlan::Unnest?

To create LogicalPlan::Unnest, it seems we can only do that in the optimization step. But this can't avoid schema change, due to `name` unnest([1,2,3]) -> [1,2,3] and `type` list(i64) -> i64.

**Comment by jayzhan211** [comment]
The Unnest function is now processing in the planning step, Expr::Unnest to LogicalPlan::Unnest. Nothing to do in the optimize phase and only the existing UnnestExec is processed in the PhysicalPlan phase. Not sure if this is a good design but it works :)

Maybe we can review the overall design, and then I will address the comments and resolve conflicts.

**Comment by alamb** [comment]
I will try and review this more carefully later today and provide feedback. Thank you @jayzhan211

File: datafusion/expr/src/expr.rs [hunk 13]
```diff
@@ -970,6 +987,43 @@ impl Expr {
...
+                args,
+            }) => {
+                let flatten_args: Vec<Expr> =
+                    args.iter().flat_map(Self::flatten_internal).collect();
+                Ok(Self::ScalarFunction(ScalarFunction {
+                    fun: built_in_function::BuiltinScalarFunction::MakeArray,
+                    args: flatten_args,
+                }))
+            }
+            _ => Err(DataFusionError::Internal(format!(
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/expr/src/expr.rs:1078]: The updated version is in https://github.com/apache/arrow-datafusion/pull/7461/files

File: datafusion/expr/src/expr_schema.rs [hunk 14]
```diff
@@ -88,6 +88,32 @@ impl ExprSchemable for Expr {
...
                 Ok((fun.return_type)(&data_types)?.as_ref().clone())
             }
+            Expr::Unnest(Unnest { array_exprs, .. }) => {
+                let data_types = array_exprs
+                    .iter()
+                    .map(|e| e.get_type(schema))
+                    .collect::<Result<Vec<_>>>()?;
+
+                if data_types.is_empty() {
+                    return Err(DataFusionError::NotImplemented(
```
Comment (jayzhan211) [inline datafusion/expr/src/expr_schema.rs:98]: TODO: use macro

File: datafusion/expr/src/logical_plan/builder.rs [hunk 15]
```diff
@@ -1038,6 +1042,114 @@ impl LogicalPlanBuilder {
     pub fn unnest_column(self, column: impl Into<Column>) -> Result<Self> {
         Ok(Self::from(unnest(self.plan, column.into())?))
     }
+
+    pub fn join_unnest_plans(
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/expr/src/logical_plan/builder.rs:1089]: We need `join` function for each unnested column. Otherwise, we get something like (1, 4), (1, 5), (2, 4), (2, 5), (3, 4), (3, 5), (3, null).

**Comment by alamb** [comment]
Marking this one as draft as it has bitrotted for a while and I think we need to work on simplifing the array code we have before we add more

**Comment by alamb** [comment]
I think DataFusion now supports UNNEST -- is this PR still being worked on?

**Comment by jayzhan211** [comment]
> I think DataFusion now supports UNNEST -- is this PR still being worked on?

We support single columns only, the challenging one multiple columns has not yet been supported, I remember @jonahgao is working on it. I forgot which PR you mentioned.

**Comment by jonahgao** [comment]
> > I think DataFusion now supports UNNEST -- is this PR still being worked on?
> 
> We support single columns only, the challenging one multiple columns has not yet been supported, I remember @jonahgao is working on it. I forgot which PR you mentioned.

Not started yet. I just discovered that it hasn't been implemented in https://github.com/apache/arrow-datafusion/pull/9592#discussion_r1522777692,  but I can start working on it from now on.
